pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py

```diff
@@ -247,32 +252,32 @@ def on_epoch_start(self) -> None:
     def on_batch_start(self) -> None:
         self._epoch_end_reached = False
 
-    def epoch_end_reached(self):
+    def epoch_end_reached(self) -> None:
         self._epoch_end_reached = True
         self._batch_idx = None
         self._split_idx = None
 
     def on_epoch_end(self) -> None:
         assert self._epoch_end_reached
         metrics = self.metrics
-        self._progress_bar_metrics.update(metrics[MetricSource.PBAR])
-        self._callback_metrics.update(metrics[MetricSource.CALLBACK])
-        self._logged_metrics.update(metrics[MetricSource.LOG])
```

Comment:
Removing `MetricSource` usage due to a mypy bug: https://github.com/python/mypy/issues/8186

Will remove the Enum in the follow-up PR

Reply:
Could we use a data class instead, I think using string is quite error prone ?

Reply:
If you check [3/3], you'll see metrics are typed as a `TypedDict`. A `TypedDict` is an in-between between dataclass and just `dict`. It gets converted to `dict` at runtime but we get the typing benefits.

What's cool is that `mypy` will complain if you try to access with a key string different to the ones in the `TypedDict`, so the errors are avoided!